metric_learn/sdml.py

```diff
@@ -52,24 +55,50 @@ def __init__(self, balance_param=0.5, sparsity_param=0.01, use_cov=True,
     super(_BaseSDML, self).__init__(preprocessor)
 
   def _fit(self, pairs, y):
+    if not has_installed_skggm():
+      msg = ("Warning, skggm is not installed, so SDML will use "
+             "scikit-learn's graphical_lasso method. It can fail to converge"
+             "on some non SPD matrices where skggm would converge. If so, "
+             "try to install skggm. (see the README.md for the right "
+             "version.)")
```

Comment:
Perhaps we can catch the case where scikit-learn's version fails and emit the warning then?

Reply:
I agree, here is the new version I'll commit. When using scikit-learn's graphical lasso, we try it and if an error is returned or the result is not finite, we raise a warning that will be printed before the error (if there is an error), or before returning M (if there is no error but there are NaNs) : Tell me what you think 
```python
    if HAS_SKGGM:
      theta0 = pinvh(sigma0)
      M, _, _, _, _, _ = quic(emp_cov, lam=self.sparsity_param,
                              msg=self.verbose,
                              Theta0=theta0, Sigma0=sigma0)
    else:
      try:
        _, M = graphical_lasso(emp_cov, alpha=self.sparsity_param,
                               verbose=self.verbose,
                               cov_init=sigma0)
        error = None
      except FloatingPointError as e:
        error = e
      if not np.isfinite(M).all() or error is not None:
        msg = ("Scikit-learn's graphical lasso has failed to converge. "
               "Package skggm's graphical lasso can sometimes converge on "
               "non SPD cases where scikit-learn's graphical lasso fails to "
               "converge. Try to install skggm and rerun the algorithm. (see "
               "the README.md for the right version.)")
        warnings.warn(msg)
        if error is not None:
          raise(error)  
```

Reply:
Actually, trying to come up with the right tests, I realized the following: 
in pathological diagonal cases (see above), scikit-learn's graphical lasso can return no error and  return a result that is not SPD (in the below example, a diagonal matrix with a negative coefficient on the diagonal). But then a NAN will appear when taking the `transformer_from_metric`. So maybe if using scikit-learn, I should additionally check that the result is indeed SPD and if not raise the error "SDML has failed to converge to an SPD matrix using scikit-learn's graphical lasso"
I don't know if skggm guarantees to return an SPD matrix, so maybe I should do the check there too ? (and then later if we dig deeper into that we could remove this if the test is not/[no more] needed)

Example (go in debug mode or put a print statement in SDML to see the result of the graphical lasso)



```python
    from metric_learn import SDML
    import numpy as np

    pairs = np.array([[[-10., 0.], [10., 0.]], [[0., 50.], [0., -60]]])
    y_pairs = [1, -1]

    sdml = SDML(use_cov=False, balance_param=100,verbose=True)

    diff = pairs[:, 0] - pairs[:, 1]
    emp_cov = np.identity(pairs.shape[2]) + 100 * (diff.T * y_pairs).dot(diff)

    print(emp_cov)

    sdml.fit(pairs, y_pairs)
    print(sdml.get_mahalanobis_matrix())
```

Returns: 
```
[[   40001.        0.]
 [       0. -1209999.]]
SDML will use scikit-learn's graphical lasso solver.
/home/will/Code/metric-learn/metric_learn/sdml.py:92: ConvergenceWarning: Warning, the input matrix of graphical lasso is not positive semi-definite (PSD). The algorithm may diverge, and lead to degenerate solutions. To prevent that, try to decrease the balance parameter `balance_param` and/or to set use_covariance=False.
  ConvergenceWarning)
[graphical_lasso] Iteration   0, cost  inf, dual gap 0.000e+00
/home/will/Code/metric-learn/metric_learn/_util.py:346: RuntimeWarning: invalid value encountered in sqrt
  return np.sqrt(metric)
[[2.4999375e-05           nan]
 [          nan           nan]]
```

And if we print the result of graphical lasso (note that it's the inverse of the initial matrix): 

```
[[ 2.49993750e-05  0.00000000e+00]
 [ 0.00000000e+00 -8.26446964e-07]]
```

Reply:
How about this:
1- first check if the matrix given as input is PSD. If not, show the warning that the algorithm may diverge etc
2- regardless of the solver: if an error is thrown by the solver, or the solution is not a PSD matrix, show error and abort (as there is no result or the result does not correspond to a valid distance)
3- if we fall in case 2- and we use sklearn solver, say that the user could try to install skggm as its graphical lasso solver is more stable

Reply:
I agree, I'll go for something like this: 

```python
    try:
      if HAS_SKGGM:
        theta0 = pinvh(sigma0)
        M, _, _, _, _, _ = quic(emp_cov, lam=self.sparsity_param,
                                msg=self.verbose,
                                Theta0=theta0, Sigma0=sigma0)
      else:
        _, M = graphical_lasso(emp_cov, alpha=self.sparsity_param,
                               verbose=self.verbose,
                               cov_init=sigma0)
      raised_error = None
      
    except Exception as e:
      raised_error = e
    w, V = np.linalg.eigh(M)
    if raised_error is not None or any(w < 0.):
      msg = ("There was a problem in SDML when using {}'s graphical "
             "lasso.").format("skggm" if HAS_SKGGM else "scikit-learn")
      if not HAS_SKGGM:
        skggm_advice = ("skggm's graphical lasso can sometimes converge "
                        "on non SPD cases where scikit-learn's graphical "
                        "lasso fails to converge. Try to install skggm and "
                        "rerun the algorithm. (See the README.md for the " 
                        "right version of skggm.)")
        msg += skggm_advice
      if raised_error is not None:
        msg += "The following error message was thrown: {}.".format(
          raised_error)
      raise(msg)
```

Reply:
I'll add infinite values that can be returned by skggm as a failure case too

Reply:
Done, in commit https://github.com/metric-learn/metric-learn/pull/162/commits/eb9571953a0001d54c2f4a1df47ab1519347f809